Add paper references and citations to paper view

Slartibartfass2 opened the pull request:
Closes #42 

## What I have made

- `PaperListEntry` can now also take just a `Paper` for when review data isn't shown
- `NamedList` can now take a `preListContent` snippet, which is displayed below the title but above the list itself
  - this is used for the search bar, but may also be used for descriptions or other information
- added `ReferencesAndCitationsCardContent` that displays references and citations in the paper view
  - both lists are searchable and scrollable in their respective half of the card

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (no logic added that needs unit tests)
  - [x] integration tests
  - ~~[ ] end-to-end tests~~ (not yet available)
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add references and citations card

Slartibartfass2 committed:
feat: add search bar to references lists

Slartibartfass2 committed:
refactor: make PaperListEntry also accept Paper objects

Previously it only accepted Project_Paper objects, but when the review info isn't shown, a Paper object is sufficient.

Slartibartfass2 committed:
feat: pass referenced papers from page load to component props

Slartibartfass2 committed:
refactor: project paper view now loads project paper to also use review data

Slartibartfass2 committed:
test: add tests for ReferencesAndCitationsCardContent

Slartibartfass2 committed:
refactor: export component props

Merseleo commented on src/lib/components/composites/list/NamedList.svelte:
```diff
@@ -49,14 +51,18 @@ If the option showNumberOfListItems is set to true (default: false),
 <div class="flex flex-col h-full w-full px-5 gap-y-5 overflow-hidden">
     {#await items}
         <h2>{listName}</h2>
+        {@render preListContent?.()}
```
Is it clever to always render the preListContent, even while loading the list data or in the error case? What do you think of the idea of only rendering this content, if the list could be successfully be loaded? Only want to think about this idea, but we can also drop it.

Slartibartfass2 replied:
I thought it would make sense to keep it visually consistent, and the search bar shouldn't pop out of nowhere when the list is finally loaded. IMO, we can hide it in the error case, but I would keep it while loading.

Merseleo replied:
Ok, i understand your reasoning and your suggestions seems to be a good idea: keep it while loading and hide it, when an error occurred, so that only the error is visible. 
Are you implementing this in this PR?

Slartibartfass2 replied:
Yes, I added it.

Merseleo commented on src/lib/components/composites/paper-components/PaperListEntry.svelte:
```diff
@@ -11,12 +11,19 @@
+                ? `/project/${projectId}/paper/${paperId}`
+                : `/paper/${paperId}`;
+        goto(paperLink);
+    };
```
Very thoughtful thanks, I hadn't even thought of that.

Slartibartfass2 replied:
Having a project independent paper view further complicates everything.

Merseleo replied:
I agree, but especially for the reading list etc. it makes sense, so imo we have the time to deal with this independent view and it could be helpful, so we should take on this additional effort.

Merseleo commented on src/lib/components/composites/paper-components/PaperListEntry.svelte:
```diff
@@ -11,12 +11,19 @@
 
-    const navigateToPaperView = () => goto(`/project/${projectId}/paper/${projectPaper.id}`);
+    const navigateToPaperView = () => {
+        const paperId = isProjectPaper(paper) ? paper.paper!.id : paper.id;
```
One idea (could be reused when creating the PaperInfo as well): What do you think of converting `isProjectPaper` to `asPaper` which gets a paper or project paper as now and return the paper either directly or extract it from the project paper? Maybe this could make the code more readable?

Slartibartfass2 replied:
Great catch, didn't see that.

I didn't find a way to use [extension methods](https://www.geeksforgeeks.org/how-to-use-extension-methods-in-typescript/) because there exists not `prototype` property on the generated types. So we have to pass the objects as arguments :(.

How would you reuse that in `PaperInfo`?

Merseleo replied:
I would reuse this method, as you already did when creating the `PaperInfo` in the `PaperListEntry` component.

And imo it is not a problem to pass the objects as arguments.

Slartibartfass2 replied:
Yes, I would've just preferred to call `projectPaper.asPaper()` instead of `asPaper(projectPaper)` but that's just syntactic sugar.

Merseleo commented on src/lib/components/composites/paper-components/PaperListEntry.svelte:
```diff
@@ -11,12 +11,19 @@
+        paper,
         projectId,
-        showReviewStatus = false,
+        showReviewStatus,
```
Why did you delete this? Is the default automatically `false`?

Slartibartfass2 replied:
No, this is due to the union type `PaperListEntryInterface`. Now, that `showReviewStatus` is either `boolean` or `false`, it doesn't work with default values, and we have explicitly set the values to have the type safety.

Merseleo replied:
Ok.

Merseleo commented on src/lib/model/general.ts:
```diff
@@ -18,10 +19,21 @@ interface ProjectListEntryInterface {
-    projectPaper: Project_Paper;
-    projectId: string;
-    showReviewStatus?: boolean;
+type PaperListEntryInterface =
```
Maybe it could be helpful to have a comment, why this type has two different interfaces ? 
Something like "...the interface for either a paper or a project paper?". Or do you think this would be confusing?

Merseleo replied:
But I like the renaming to paper (instead of my projectPaper)

Slartibartfass2 replied:
Good catch. I added an explanation.

Merseleo commented on src/lib/components/composites/search-bar/SearchBar.svelte:
```diff
@@ -5,13 +5,13 @@
     }
-    const { placeholderText = "Search", onSearch }: Props = $props();
+
+    const { placeholderText = "Search", onSearch, timeoutInMs = 500 }: Props = $props();
```
If you make it a prop (good idea btw), then it would be helpful to add in the usage example below.

Slartibartfass2 replied:
I agree, I added it.

Merseleo commented on tests/integration/paper-components/paper-list-entry.test.ts:
```diff
@@ -91,14 +86,8 @@ describe("PaperListEntryComponent", () => {
-                projectId: "0",
+                paper: Papers.demoPaper1,
+                projectId: undefined,
                 showReviewStatus: false,
```
Actually this test should test, that no review information are shown although it is a project paper and information exist. So I think you should actually not provide a paper, but a project paper.

Slartibartfass2 replied:
I'm a bit confused, the test you reference tests a custom `onClick` function. Which one, do you mean?

Merseleo replied:
Ops, sorry. I was confused by the way Github showed me the diffed and it seems as you deleted code in other tests. My mistake ...

Merseleo commented on tests/integration/paper-components/paper-view/cards/references-and-citations-card-content.test.ts:
```diff
@@ -0,0 +1,77 @@
+        expect(searchBars).toHaveLength(2);
+    });
+
+    test("When references are searched, then only search results are shown", async () => {
```
```suggestion
    test("When references are searched, then only matching reference papers are shown", async () => {
```

Merseleo commented on tests/integration/paper-components/paper-view/cards/references-and-citations-card-content.test.ts:
```diff
@@ -0,0 +1,77 @@
+            },
+        });
+
+        expect(Papers.demoPaper1.title).not.toEqual(Papers.demoPaper2.title);
```
This is a use case for assert, isn't it?

Slartibartfass2 replied:
It's the same, isn't it?

Merseleo replied:
It is, but if I remember correctly we once discussed about where using assert could make sense (especially in case of readability). Yes afaik it is the same and you dont need to change anything.

Merseleo commented on tests/integration/paper-components/paper-view/cards/references-and-citations-card-content.test.ts:
```diff
@@ -0,0 +1,77 @@
+        expect(paper2).not.toBeInTheDocument();
+    });
+
+    test("When citations are searched, then only search results are shown", async () => {
```
```suggestion
    test("When citations are searched, then only matching citation papers are shown", async () => {
```

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/PaperReviewCard.svelte:
```diff
@@ -0,0 +1,41 @@
+
+<!--
+@component
+Paper Card for paper review information in the Paper View component.
```
```suggestion
Card for review information of a paper in the Paper View component.
```

Slartibartfass2 replied:
Changed it.

Merseleo replied:
I think you forgot to merge this update as well? Because the comment in `PaperResearchContextCard` is still the old one and not correct anymore for the new component.

Slartibartfass2 replied:
Currently, it's "Paper Card for review information of a paper in the Paper View component." (I kept the "Paper ")
Is that alright?

Merseleo replied:
No it's not, as this is not only the card only for review information but also for the references, isn't it?

Slartibartfass2 replied:
I improved the documentation.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/PaperReviewCard.svelte:
```diff
@@ -0,0 +1,41 @@
+
+    const tabs = [
+        { value: "1", label: "Review Criteria" },
+        { value: "2", label: "References/Citations" },
```
Imo it is not so user friendly, that the tabs are not only completely renamed ("Review Citeratia" to "Review Information" makes totally sense, but why "References/Citations" and "Forward/Backward References") but also the order of the tabs is changed?! My suggestion is to only rename "criteria" to "information" and change the default tab, but keep the order. What do you think?

Slartibartfass2 replied:
Probably, forgot to rename "References/Citations" in the `PaperReviewCard`, I changed that.
We decided to set this order, so that in the review mode the review information is seen directly and the other mode not, so the user has to actively decide to see the review information, which might bias them.

Merseleo replied:
I completely understand your point, but why not change the default tab instead of the order of the tabs. So that always "References" is the first tab und "Review information" the second, but when I enter the review mode, the second tab "review information" is automatically selected?

Slartibartfass2 replied:
As discussed, we keep it this way for now and discuss it later when we gather some user feedback. I mentioned it [here](https://github.com/SE-UUlm/snowballr-frontend/issues/211).

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/PaperReviewCard.svelte:
I do not really understand, why you have this and the PaperReaearchContextCard wrapper components? They seem to be very very similar and the advantage of this is not quite clear to me yet. Why don't have on component and change one tab name and the default tab depending on the review mode (which will be stored in a store soon and can be easily accessed) and the change the child component depending on the review mode? Because all the cards make it a bit confusing imo.

Slartibartfass2 replied:
I agree, they differ mostly on the review criteria tab, which was the reason for me to do it separately.
We can merge them.

Merseleo replied:
Ok, thank you.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+    import PaperListEntrySkeleton from "../../PaperListEntrySkeleton.svelte";
+    import type { Paper } from "$lib/model/api/paper";
+
+    export interface ReferencesAndCitationsCardContentProps {
+        backwardReferencedPapers: Promise<Paper[]>;
+        forwardReferencedPapers: Promise<Paper[]>;
+    }
```
Why don't you move this interface in the general.ts, if you want it to use somewhere else?

Slartibartfass2 replied:
I think it makes sense to keep it with the component because if it wouldn't be in the same file, one would have to go to another file to see the props, which seems wrong/confusing to me. This way, we see the props directly and can use it in components that use this component.
Furthermore, changes can be directly applied in here and not somewhere else.

Merseleo replied:
Ok, if it works this way, I am conform with this (because WebStorm show errors because of the export of the interface and the way you did causes problems in WebStorm, but nothing important).

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+
+    function filterPapers(allPapers: Paper[], searchText: string) {
+        const fzf = new Fzf(allPapers, {
+            selector: (paper) => `#${paper!.id} ${paper!.title}`,
```
What do you think of searching by more keywords than only the id and paper title? So by authors or external id or so too? And maybe we could outsource the different search functions as we need to reuse this function in the general paper search too later.

Merseleo replied:
And if you outsource this function, imo it would be helpful to have a comment for this function which describes which keywords or content of the paper is used for search.

Slartibartfass2 replied:
I outsourced the search function and we now also check the authors. I wouldn't add more because the user only sees these properties. Otherwise, it would be confusing if you type the publication name or something but don't see that it matches it.

Merseleo replied:
Makes sense, I am fine with your solution.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+</script>
+
+<section class="flex flex-col h-full gap-5">
+    <div class="flex h-full overflow-hidden">
```
Why did you encapsulated the named list in these divs? I delete them and it seems as they make no difference?

Slartibartfass2 replied:
The divs are important to limit the overflow of the lists.

Merseleo replied:
I tested it again and combined with your new `flex-[1_1_0]` it seems to work now, because before your change, the divs didn't make a difference, when I tested it, but now. Thank you for fixing it.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+            numberOfSkeletons={3}
+        >
+            {#snippet preListContent()}
+                <SearchBar onSearch={filterBackwardReferencedPapers} timeoutInMs={0} />
```
Why did you override the timeout to 0. I thought we tested it and agreed, that 500ms is ok?

Slartibartfass2 replied:
We did that for asynchronous calls, but in this case, we have all the data on the client (maybe we should change that later) and therefore can update the list immediately.

Merseleo replied:
Ah ok, that makes sense. I didn't consider, that all data are already in the client.

Slartibartfass2 replied:
Yes, maybe later we want to do the filtering on the server side, and then we would use a delay instead again.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+                <SearchBar onSearch={filterBackwardReferencedPapers} timeoutInMs={0} />
+            {/snippet}
+            {#snippet listItemComponent(paper)}
+                <PaperListEntry {paper} projectId={undefined} showReviewStatus={false} />
```
This seems a bit weird to me: why you always set the projectId to undefined and the review status to false? I think there should be the option to have a list of project papers in there?

Slartibartfass2 replied:
This is the same answer as for https://github.com/SE-UUlm/snowballr-frontend/pull/184/files#r1957159395.

When using a union type, we have to explicitly set the values so that the type can be inferred correctly.
When we request the backward and forward referenced papers, we only receive `Paper` object, which is why we can only pass them as `Paper` object and not as `Project_Paper` objects here.

Merseleo replied:
Ok, in my head I was still thinking you solved it with optional parameters, but I had another look at your interface and now it makes sense to me.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+<section class="flex flex-col h-full gap-5">
+    <div class="flex h-full overflow-hidden">
+        <NamedList
+            listName="References"
+            items={backwardReferencedPapers}
+            showNumberOfListItems
+            numberOfSkeletons={3}
```
```suggestion
            listName="References"
            items={backwardReferencedPapers}
            showNumberOfListItems
            numberOfSkeletons={3}
            emptyHint="No references found."
```

Slartibartfass2 replied:
Added it.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReferencesAndCitationsCardContent.svelte:
```diff
@@ -0,0 +1,80 @@
+    </div>
+    <div class="flex h-full overflow-hidden">
+        <NamedList
+            listName="Citations"
+            items={forwardReferencedPapers}
+            showNumberOfListItems
+            numberOfSkeletons={3}
```
```suggestion
            listName="Citations"
            items={forwardReferencedPapers}
            showNumberOfListItems
            numberOfSkeletons={3}
            emptyHint="No citations found.
```

Slartibartfass2 replied:
Added it.

Merseleo commented on src/routes/paper/[paperId]/+page.ts:
I think you forgot the error handling (so your catch cases). Because if the paper (id) was not found or the server is not running and anything could be loaded, the plain grpc errors are shown to the user and no custom messages which could be helpful or are understandable.
<img width="759" alt="Bildschirmfoto 2025-02-15 um 17 03 27" src="https://github.com/user-attachments/assets/f6d5ff10-010c-4be2-9084-4d2eac35430f" />
<img width="742" alt="Bildschirmfoto 2025-02-15 um 17 52 30" src="https://github.com/user-attachments/assets/ae331503-b5f8-4c60-97c0-7189573819b4" />

So I think you need to catch the errors and forward better error messages as on the homepage.

Slartibartfass2 replied:
Yes, that was missing in `NamedList`. I changed it.

Merseleo replied:
Perfect. I tested it and it works 👍.
Would you mind testing the whole behaviour (i.e. whether the hint or error is shown when the map is rendered with an empty list or an error). I know it doesn't increase the coverage, but it feels a bit more complete or do you want to leave it out because the list itself is already tested (which I would totally understand)

Slartibartfass2 replied:
No, I think that's fine. I added them.

Merseleo replied:
Thanks.

Merseleo reviewed (changes requested):
It seems as if we are making great strides towards a finished paper page. 
I think, on the whole, the two list work greatly. I have also tested it and if there is data, it works as intended, only if there is no data I don't see such good error messages as a user. 
Another thing that confused me: in the mockups we have a button to add a new paper in the normal view after the list or if there were no search results. Did you leave this out on purpose and if so, then it would be good to create an issue for it or link it and record it somewhere in the code.

Slartibartfass2 committed:
feat: apply suggestions from code review

Co-authored-by: Merseleo <43602965+Merseleo@users.noreply.github.com>

Slartibartfass2 reviewed (commented):
Thanks for the review.

Slartibartfass2 committed:
refactor: use `asPaper` instead of `isProjectPaper`

This removes the duplicated condititions to get the actual paper object.

Slartibartfass2 committed:
refactor: merge `PaperResearchContextCard` and `PaperReviewCard`

They only differ in some minor conditions, so we can easily merge them and put the conditions inside.

Slartibartfass2 committed:
refactor: extract filter methods to separate file

Slartibartfass2 committed:
docs: add some missing documentation

Slartibartfass2 committed:
fix: fix paper card overflow

Slartibartfass2 commented:
> It seems as if we are making great strides towards a finished paper page. I think, on the whole, the two list work greatly. I have also tested it and if there is data, it works as intended, only if there is no data I don't see such good error messages as a user. Another thing that confused me: in the mockups we have a button to add a new paper in the normal view after the list or if there were no search results. Did you leave this out on purpose and if so, then it would be good to create an issue for it or link it and record it somewhere in the code.

For the feature you ask for, [this issue](https://github.com/SE-UUlm/snowballr-frontend/issues/43) already exists.

Merseleo commented on src/lib/model/general.ts:
```diff
@@ -18,10 +19,32 @@ interface ProjectListEntryInterface {
 }
 
 export type { ValidationResult, ApiError, ProjectListEntryInterface, PaperListEntryInterface };
+export { isProjectPaper, asPaper };
```
But now can keep the isProjectPaper method locally as you already export the asPaper method and don't use the isProjectPaper method anymore?

Slartibartfass2 replied:
Yes, I kept it exported if we may ever want to use it. But if so, we can export it again.

Merseleo replied:
Basically yes, but I just think that you should export as little as possible and thus expose, especially because, as you say, if you then need it, you can simply export it quickly.

Merseleo commented on src/lib/components/composites/search-bar/SearchBar.svelte:
```diff
@@ -54,9 +54,12 @@ and the search method is triggered with an empty string,
 by the search.
 
+To limit the amount of calls (if used with backend calls), the search only starts after a
+specified timout in which no other events occur. The default for this is 500 ms.
```
```suggestion
To limit the amount of calls (if used with backend calls), the search only starts after a
specified timeout in which no other events occur. The default timeout is 500 ms.
```

Slartibartfass2 replied:
Definitely need spell checking in VS Code. I changed it.

Merseleo replied:
A great idea ;).

Merseleo commented on src/lib/components/composites/list/NamedList.svelte:
```diff
@@ -12,6 +12,8 @@
         showNumberOfListItems?: boolean;
         numberOfItems?: number;
         emptyHint?: string;
+        errorHint?: string;
```
I already solved this problem actually by catching the errors in the page load and rename the error message, so dont need to think of setting an error hint, especially as your way only allows to display one error message and not multiple, which is not ideal imo.

Slartibartfass2 replied:
If the error hint is not set, it displays the content of the error object, so both ways work.

Merseleo replied:
Ok I am also fine with that.

Merseleo commented on src/lib/components/composites/paper-components/PaperListEntry.svelte:
```diff
@@ -11,12 +11,19 @@
+            projectId !== undefined
+                ? `/project/${projectId}/paper/${paperId}`
+                : `/paper/${paperId}`;
+        goto(paperLink);
```
I found another problem with this goto (didn't recognize this earlier, sorry): Try to add two papers as references and click on one reference, which should navigate you to this new page. It does happen, but the title in the navigation bar and everything else will not be reloaded and the old paper information are displayed.

Slartibartfass2 replied:
Yes, I already noticed this too, and I'm convicted this is some svelte behavior that we don't know because the URL changes, but the page stays the same. I would like to handle this in a separate issue, e.g., when we face navigating papers or in an entirely new one.

Merseleo replied:
Ok, then I write a bug issue for this problem.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/PaperResearchContextCard.svelte:
```diff
@@ -1,37 +1,73 @@
 <!--
 @component
-Paper Card for paper research context in the Paper View component.
+Paper Card for review information of a paper in the Paper View component.
```
```suggestion
Card for displaying the review information and forward / backward references of a paper in the Paper View component.
```

Slartibartfass2 replied:
Changed it.

Merseleo replied:
Better now imo.

Merseleo commented on src/lib/filters.ts:
Did you test these methods or it is ok to not explicitly test them?
And second question: do you think we should move this file in the `utils` folder?

Slartibartfass2 replied:
I am not sure if we should, actually. I tested it through the component because this way I see if it works for the component. The method itself is just FZF, so IDK....

Merseleo replied:
I agree, I think it is ok to "only" implicitly test them and fzf should work correctly...

Merseleo commented on src/lib/filters.ts:
```diff
@@ -0,0 +1,59 @@
+ * @param searchText - Search text
+ * @returns List of items that match the search text
+ */
+function filter<T>(items: T[], selector: (item: T) => string, searchText: string) {
+    // Here we case the items to string[] and selector to Selector<string> because we can't just use a selector
+    // with a generic type. This is weird but it's the only way to make it work.
+    const fzf = new Fzf(items as string[], {
+        selector: selector as Selector<string>,
+        casing: "case-insensitive",
+    });
+    return fzf.find(searchText).map((result) => result.item) as T[];
+}
```
Nice typescript generics, very useful for my exam preparation ;) (close it by yourself)

Merseleo reviewed (changes requested):
Great work, I think we will finish it in this round.

Slartibartfass2 committed:
test: add more tests for ReferencesAndCitationsCardContent

Slartibartfass2 committed:
chore: remove preListContent in error state of NamedList

Slartibartfass2 reviewed (commented):
This should be all. Thanks for the review.

Merseleo reviewed (changes requested):
I think you missed some comments.

Slartibartfass2 committed:
docs: add some documentation

Slartibartfass2 committed:
refactor: move filters.ts to utils directory

Slartibartfass2 reviewed (commented):
This time, it should be all answered.

Merseleo reviewed (approved):
I think all problems are solved, everything well tested and documented and the functionality should be added to our develop version ...
